Tutorial - Android session replays (#8933)
PostHog/posthog.com: contents/tutorials/android-session-replays.md

---
title: How to set up Android session replays
date: 2024-07-10
author:
  - lior-neu-ner
tags:
  - session replay
---

[Session replays](/session-replay) are a useful support tool for understanding how users are interacting with your Android app. It also helps you debug and recreate issues. 

> **

Reviewer: Whats the intention here? seems like a typo somewhere, its not breaking a line at least

Author: ```suggestion

```

  imageDark={"https://res.cloudinary.com/dmukukwp6/image/upload/v1720608972/posthog.com/contents/Screenshot_2024-07-10_at_11.55.29_AM.png"} 
  alt="Android session replays in PostHog" 
  classes="rounded"
/>

## 3. (Optional) Mask sensitive data

Your replays may contain sensitive information. For example, if you're building a banking app you may not want to capture how much money a user has in their account.

To replace any type of `View` with a redacted version in the replay, set the [tag](https://developer.android.com/reference/android/view/View#tags) to `ph-no-capture`.

In the below code, we do this for the email and password fields:

```kotlin file=MainActivity.kt
// your existing imports
import androidx.compose.ui.platform.testTag

// your existing of your code

@Composable
fun WelcomeScreen(navController: NavController, modifier: Modifier = Modifier) {
    Surface(modifier = Modifier.fillMaxSize()) {
        Column(
            modifier = Modifier.fillMaxSize(),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.Center
        ) {
            Text(
                text = "Welcome!",

>                 modifier = Modifier.testTag("ph-no-capture") // Add this modifier to blur the text in replays

Reviewer: ```suggestion
                modifier = Modifier.testTag("ph-no-capture") // Add this modifier to redact the text in replays
```

>             )

Author: @marandaneto this approach didnt work for me. The Text still appears as unblurred in the replays. Any ideas on how to do this?

Reviewer: mm I have to check, SwiftUI did require some extra stuff to make it work, I will check

Reviewer: https://github.com/PostHog/posthog.com/pull/9049
a limitation sadly, will investigate deeper.

Author: Thanks! Is there an example of a component you can suggest I can use this on? Would love to show how to redact in this tutorial

Reviewer: In this case, I'd not use Jetpack compose then until I figure this out.

Author: I see. Whats the alternative to Jetpack compose that i should use?

Reviewer: just the standard declarative views using xml
https://developer.android.com/develop/ui/views/layout/declaring-layout

        android:textSize="24sp"
        android:gravity="center" />

    <Button
        android:id="@+id/logoutButton"
        android:layout_width="match_parent"
        android:layout_height="wrap_content"
        android:text="Logout" />

</LinearLayout>
```

Next we create our activities. In `java/com.example.android_session_replays`, create two new Kotlin files `LoginActivity.kt` and `WelcomeActivity.kt`. Add the following code to each file:

```kotlin file=LoginActivity.kt
package com.example.android_session_replays

import android.content.Intent
import android.os.Bundle
import android.widget.Button
import android.widget.EditText
import androidx.appcompat.app.AppCompatActivity

class LoginActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)

        val email = findViewById<EditText>(R.id.email)
        val password = findViewById<EditText>(R.id.password)
        val loginButton = findViewById<Button>(R.id.loginButton)

        loginButton.setOnClickListener {
            val intent = Intent(this, WelcomeActivity::class.java)

>             startActivity(intent)

Reviewer: ```suggestion
            startActivity(intent)
            finish()
```

otherwise, the current screen is still open and in memory

        <!-- ... rest of the config -->
    </application>
</manifest>
```

To check your setup, build and run your app a few times. Enter in any values in the textfields and click the **Log in** button. You should start session replays in the [Session replay tab](https://us.posthog.com/replay/recent) in PostHog 🎉

<ProductScreenshot
  imageLight={"https://res.cloudinary.com/dmukukwp6/image/upload/v1722266411/posthog.com/contents/Screenshot_2024-07-29_at_4.19.27_PM.png"} 
  imageDark={"https://res.cloudinary.com/dmukukwp6/image/upload/v1722266410/posthog.com/contents/Screenshot_2024-07-29_at_4.19.59_PM.png"} 
  alt="Android session replays in PostHog" 
  classes="rounded"
/>

## 3. (Optional) Mask sensitive data

Your replays may contain sensitive information. For example, if you're building a banking app you may not want to capture how much money a user has in their account. 

To replace any type of `View` with a redacted version in the replay, set the [tag](https://developer.android.com/reference/android/view/View#tags) to `ph-no-capture`.

The example below illustrates how to do this for the **Welcome** text in the second screen:

> ```kotlin file=activity_welcome.xml

Reviewer: ```suggestion
```xml file=activity_welcome.xml
```

---
title: How to set up Android session replays
date: 2024-07-10
author:
  - lior-neu-ner
tags:
  - session replay
---

[Session replays](/session-replay) are a useful support tool for understanding how users are interacting with your Android app. It also helps you debug and recreate issues. 
To show how you to set it up with PostHog, in this tutorial we create a basic Kotlin app, add PostHog, and [enable session recordings](/docs/session-replay/mobile#android).

## 1. Create a basic Android app

Our sample app will have two screens:

- The first screen is be a `login` screen with email and password textfields.
- The second screen is a simple screen with welcome text and logout button.

> The first step is to create a new app. Open [Android Studio](https://developer.android.com/studio) and create a new project. Select `No Activity`, name your project `Android-Session-Replays`, and use the defaults for everything else.

Reviewer 2: It is fine to have the project be named `Android-Session-Replays` so you don't have to change everything, I just think the title and slug should be singular.

> Then, navigate to the `res` directory and create a new folder `layout` in it. In `res/layout`, create two new layout resource files `activity_login.xml` and `activity_welcome.xml`. Add the following code to each file:

Reviewer 2: ```suggestion
Then, navigate to the `res` directory and create a new directory `layout` in it. In `res/layout`, create two new layout resource files `activity_login.xml` and `activity_welcome.xml`. Add the following code to each file:
```

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)

        val email = findViewById<EditText>(R.id.email)
        val password = findViewById<EditText>(R.id.password)
        val loginButton = findViewById<Button>(R.id.loginButton)

        loginButton.setOnClickListener {
            val intent = Intent(this, WelcomeActivity::class.java)
            startActivity(intent)
            finish()
        }
    }
}
```

```kotlin file=WelcomeActivity.kt
package com.example.android_session_replays

import android.content.Intent
import android.os.Bundle
import android.widget.Button
import androidx.appcompat.app.AppCompatActivity

class WelcomeActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_welcome)

        val logoutButton = findViewById<Button>(R.id.logoutButton)

        logoutButton.setOnClickListener {
            val intent = Intent(this, LoginActivity::class.java)
            startActivity(intent)
            finish()
        }
    }
}
```

> Lastly, add your new activities to `AndroidManifext.xml`:

Reviewer 2: ```suggestion
Lastly, go to the `manifests` directory and add your new activities to `AndroidManifext.xml`:
```

        android:label="@string/app_name"
        android:roundIcon="@mipmap/ic_launcher_round"
        android:supportsRtl="true"
        android:theme="@style/Theme.AndroidSessionReplays"
        tools:targetApi="31">
        <activity android:name=".WelcomeActivity"
            android:exported="true"/>
        <activity android:name=".LoginActivity"
            android:exported="true">
            <intent-filter>
                <action android:name="android.intent.action.MAIN" />
                <category android:name="android.intent.category.LAUNCHER" />
            </intent-filter>
        </activity>
    </application>

</manifest>
```

Our basic set up is now complete. Build and run your app to see it in action.

![Basic setup of the Android app](https://res.cloudinary.com/dmukukwp6/video/upload/v1722266060/posthog.com/contents/android-sample-app.mp4)

## 2. Add PostHog to your app

First, add the [PostHog Android SDK](/docs/libraries/android) as a dependency in your `Gradle Scripts/build.gradle.kts (Module: app)` file. You can find the latest version on our [GitHub](https://github.com/PostHog/posthog-android/blob/main/CHANGELOG.md). 

> Note that session replays require SDK version `3.4.0` or higher.

Reviewer 2: ```suggestion
**Note:** Session replay requires SDK version `3.4.0` or higher.
```

```gradle_kotlin file=app/build.gradle
dependencies {
    implementation("com.posthog:posthog-android:3.+")
    //... other dependencies
}
```

Sync your project with your Gradle file changes.

> Next, we create a Kotlin class where we can configure our PostHog instance. In the `src/main/java/com.example.android_session_replays` folder, add a new file `MySessionReplaysApplication.kt` and then add the following code:

Reviewer 2: My folder structure is not like that for some reason 🤔 
```suggestion
Next, we create a Kotlin class where we can configure our PostHog instance. In the `java/com.example.android_session_replays` directory, add a new file `MySessionReplaysApplication.kt` and then add the following code:
```

```kotlin file=MySessionReplaysApplication.kt
package com.example.android_session_replays

import android.app.Application
import com.posthog.android.PostHogAndroid
import com.posthog.android.PostHogAndroidConfig

class MySessionReplaysApplication : Application() {
    companion object {
        private const val POSTHOG_API_KEY = "<ph_project_api_key>"
        private const val POSTHOG_HOST = "<ph_client_api_host>" // usually 'https://us.i.posthog.com' or 'https://eu.i.posthog.com'
    }

    override fun onCreate() {
        super.onCreate()
        val config = PostHogAndroidConfig(
            apiKey = POSTHOG_API_KEY,
            host = POSTHOG_HOST 
        ).apply {
            sessionReplay = true
            sessionReplayConfig.maskAllTextInputs = false // Whether all texts are masked or redacted (default is enabled)
            sessionReplayConfig.screenshot = true
        }
        PostHogAndroid.setup(this, config)
    }
}
```

> To get your PostHog API key and host, [sign up to PostHog](https://us.posthog.com/signup). Then, you can find your API key and host in your [project settings](https://us.posthog.com/settings/project).

Reviewer 2: ```suggestion
To get your PostHog API key and host, [sign up for PostHog](https://us.posthog.com/signup). Then, you can find your API key and host in your [project settings](https://us.posthog.com/settings/project).
```

We now need to register our custom application class. Go to `app/manifests/AndroidManifest.xml` and add `android:name=".MySessionReplaysApplication"` within the `<application>` tag:

```XML file=app/manifests/AndroidManifest.xml
<?xml version="1.0" encoding="utf-8"?>
<manifest xmlns:android="http://schemas.android.com/apk/res/android"
    xmlns:tools="http://schemas.android.com/tools">
    <!-- ... rest of the config -->
    <application
        android:name=".MySessionReplaysApplication"
        <!-- ... rest of the config -->
    </application>
</manifest>
```

> To check your setup, build and run your app a few times. Enter in any values in the textfields and click the **Log in** button. You should start session replays in the [Session replay tab](https://us.posthog.com/replay/recent) in PostHog 🎉

Reviewer 2: ```suggestion
To check your setup, build and run your app a few times. Enter in any values in the text fields and click the **Log in** button. You should start session replays in the [session replay tab](https://us.posthog.com/replay/recent) in PostHog 🎉.
```

<ProductScreenshot

>   imageLight={"https://res.cloudinary.com/dmukukwp6/image/upload/v1722266411/posthog.com/contents/Screenshot_2024-07-29_at_4.19.27_PM.png"} 
>   imageDark={"https://res.cloudinary.com/dmukukwp6/image/upload/v1722266410/posthog.com/contents/Screenshot_2024-07-29_at_4.19.59_PM.png"} 

Reviewer 2: ```suggestion
  imageLight="https://res.cloudinary.com/dmukukwp6/image/upload/v1722266411/posthog.com/contents/Screenshot_2024-07-29_at_4.19.27_PM.png"
  imageDark="https://res.cloudinary.com/dmukukwp6/image/upload/v1722266410/posthog.com/contents/Screenshot_2024-07-29_at_4.19.59_PM.png"
```

Reviewer 2: This is so cool tbh

  alt="Android session replays in PostHog" 
  classes="rounded"
/>

## 3. (Optional) Mask sensitive data

> Your replays may contain sensitive information. For example, if you're building a banking app you may not want to capture how much money a user has in their account. 

Reviewer 2: ```suggestion
Your replays may contain sensitive information. For example, if you're building a banking app you may not want to capture how much money a user has in their account. PostHog tries to automatically mask sensitive data (like the password text field), but sometimes you need to do it manually.
```

To replace any type of `View` with a redacted version in the replay, set the [tag](https://developer.android.com/reference/android/view/View#tags) to `ph-no-capture`.

The example below illustrates how to do this for the **Welcome** text in the second screen:

```xml file=activity_welcome.xml
<!-- rest of your XML -->
    <TextView

>         android:tag="ph-no-capture"```

Reviewer 2: ```suggestion
        android:tag="ph-no-capture"
```

<!-- rest of your XML -->
```

> Now the welcome messages shows up in replays like this:

Reviewer 2: ```suggestion
Now, the welcome messages shows up in replays like this:
```
